src/core/dbus-service.c

```diff
@@ -161,12 +161,25 @@ static int bus_service_method_mount(sd_bus_message *message, void *userdata, sd_
         else if (!path_is_absolute(dest) || !path_is_normalized(dest))
                 return sd_bus_error_set(error, SD_BUS_ERROR_INVALID_ARGS, "Destination path must be absolute and normalized");
 
+        if (access(src, F_OK) < 0)
+                return sd_bus_error_setf(error, SD_BUS_ERROR_FILE_NOT_FOUND, "Source path '%s' does not exist", src);
```

Comment:
please propagate the errno, there might be many other reasons access() fails.

iirc sd_bus_error_set_errnof() already turns ENOENT into SD_BUS_ERROR_FILE_NOT_FOUND, hence you might just use that

Reply:
I added this for 2 reasons: first of all, without this dfuzzer crashes when calling this with a non-existent path, but ONLY on dbus-broker.service and nothing else. I haven't the faintest clue why. The early check avoids it. The second reason is that I can add a more friendly error message here, can't do that from the async child process.

I don't think this is really a TOCTOU situation? It is checked again later too, so in a sense it's checked twice. But just as the basic sanity checks above it, I think it makes sense to do basic checks immediately, and avoid forking the child process entirely, and reject the request immediately with a clear error.

Reply:
the right way to address the dbus-broker thing is to track down what's going on, not to add a work-around.

actually, given this is an entirely foreign file we are getting pointed to, I#d really prefer if we'd not touch it here from the parent. I mean we reach into certain service paths from PID 1 at various places, but i'd try to avoid it if it is likely the paths might point to completely arbitrary places (i.e. where for example it isn't known that this will point only to /run, i.e. a local tmpfs). and that's the case here. the ddis referenced here might be on nfs or whatnot. hence I'd really just access them in the child, not the parent.

hence, validate the paths "dry" ahead of time, but dereference them only in the child

Reply:
> the right way to address the dbus-broker thing is to track down what's going on, not to add a work-around.

I have no idea how. It doesn't make sense as it fails only in that case and for nothing else. Suggestions?

Reply:
Somehow, for some reason, the bus gets disconnected when the method fails when called on dbus-broker.service, but not on any other service. I have no idea what's going on. When it succeeds it's fine, there are no issues. If there are no suggestions, I have no idea what to do besides leaving this check in place.

```
(sd-mount-in-ns)[601]: Failed to resolve source path '/org/freedesktop/foo': No such file or directory
(sd-mount-in-ns)[601]: dbus-broker.service: Failed to mount '/org/freedesktop/foo' on '/org/freedesktop/foo' in unit's namespace: No such file or directory
systemd[1]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RemoveMatch cookie=387 reply_cookie=0 signature=s error-name=n/a error-message=n/a
systemd[1]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RemoveMatch cookie=388 reply_cookie=0 signature=s error-name=n/a error-message=n/a
systemd[1]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RemoveMatch cookie=389 reply_cookie=0 signature=s error-name=n/a error-message=n/a
systemd[1]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RemoveMatch cookie=390 reply_cookie=0 signature=s error-name=n/a error-message=n/a
systemd[1]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RemoveMatch cookie=391 reply_cookie=0 signature=s error-name=n/a error-message=n/a
systemd[1]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RemoveMatch cookie=392 reply_cookie=0 signature=s error-name=n/a error-message=n/a
systemd[1]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RemoveMatch cookie=393 reply_cookie=0 signature=s error-name=n/a error-message=n/a
systemd[1]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RemoveMatch cookie=394 reply_cookie=0 signature=s error-name=n/a error-message=n/a
systemd[1]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RemoveMatch cookie=395 reply_cookie=0 signature=s error-name=n/a error-message=n/a
systemd[1]: Bus bus-api-system: changing state RUNNING → CLOSED
systemd[1]: Received SIGCHLD from PID 601 ((sd-mount-in-ns).
systemd[1]: Child 601 ((sd-mount-in-ns) died (code=exited, status=1/FAILURE)
systemd[1]: init.scope: Child 601 belongs to init.scope.
systemd[1]: dbus-broker.service: Child 601 belongs to dbus-broker.service.
systemd[1]: dbus-broker.service: Control process exited, code=exited, status=1/FAILURE
systemd[1]: dbus-broker.service: Got final SIGCHLD for state mounting.
systemd[1]: dbus-broker.service: 'BindMount' method failed: org.freedesktop.DBus.Error.Failed
systemd[1]: dbus-broker.service: Changed mounting -> running
systemd[1]: varlink-124-124: Sending message: {"parameters":{"cgroups":[{"mode":"auto","path":"/system.slice/dbus-broker.service","property":"ManagedOOMSwap"},{"mode":"auto","path":"/system.slice/dbus-broker.service","property":"ManagedOOMMemoryPressure","limit":0}]},"continues":true}
systemd[1]: Bus bus-api-system: changing state UNSET → OPENING
systemd[1]: sd-bus: starting bus bus-api-system by connecting to /run/dbus/system_bus_socket...
systemd[1]: Bus bus-api-system: changing state OPENING → AUTHENTICATING
systemd[1]: Registering bus object implementation for path=/org/freedesktop/systemd1 iface=org.freedesktop.systemd1.Manager
systemd[1]: Registering bus object implementation for path=/org/freedesktop/systemd1/job iface=org.freedesktop.systemd1.Job
systemd[1]: Registering bus object implementation for path=/org/freedesktop/systemd1/unit iface=org.freedesktop.systemd1.Unit
systemd[1]: Registering bus object implementation for path=/org/freedesktop/systemd1/unit iface=org.freedesktop.systemd1.Automount
systemd[1]: Registering bus object implementation for path=/org/freedesktop/systemd1/unit iface=org.freedesktop.systemd1.Device
systemd[1]: Registering bus object implementation for path=/org/freedesktop/systemd1/unit iface=org.freedesktop.systemd1.Mount
systemd[1]: Registering bus object implementation for path=/org/freedesktop/systemd1/unit iface=org.freedesktop.systemd1.Path
systemd[1]: Registering bus object implementation for path=/org/freedesktop/systemd1/unit iface=org.freedesktop.systemd1.Scope
systemd[1]: Registering bus object implementation for path=/org/freedesktop/systemd1/unit iface=org.freedesktop.systemd1.Service
systemd[1]: Registering bus object implementation for path=/org/freedesktop/systemd1/unit iface=org.freedesktop.systemd1.Slice
systemd[1]: Registering bus object implementation for path=/org/freedesktop/systemd1/unit iface=org.freedesktop.systemd1.Socket
systemd[1]: Registering bus object implementation for path=/org/freedesktop/systemd1/unit iface=org.freedesktop.systemd1.Swap
systemd[1]: Registering bus object implementation for path=/org/freedesktop/systemd1/unit iface=org.freedesktop.systemd1.Target
systemd[1]: Registering bus object implementation for path=/org/freedesktop/systemd1/unit iface=org.freedesktop.systemd1.Timer
systemd[1]: Registering bus object implementation for path=/org/freedesktop/LogControl1 iface=org.freedesktop.LogControl1
systemd[1]: systemd-resolved.service: Watching D-Bus name 'org.freedesktop.resolve1'.
systemd[1]: systemd-networkd.service: Watching D-Bus name 'org.freedesktop.network1'.
systemd[1]: systemd-logind.service: Watching D-Bus name 'org.freedesktop.login1'.
systemd[1]: systemd-hostnamed.service: Watching D-Bus name 'org.freedesktop.hostname1'.
systemd[1]: systemd-homed.service: Watching D-Bus name 'org.freedesktop.home1'.
systemd[1]: systemd-oomd.service: Watching D-Bus name 'org.freedesktop.oom1'.
systemd[1]: systemd-timesyncd.service: Watching D-Bus name 'org.freedesktop.timesync1'.
systemd[1]: systemd-importd.service: Watching D-Bus name 'org.freedesktop.import1'.
systemd[1]: Successfully connected to API bus.
systemd[1]: Successfully connected to system bus.
systemd[1]: Bus bus-api-system: changing state AUTHENTICATING → HELLO
```

Reply:
Found it: there was another use of the service state outside of service.c in manager.c for manager_dbus_is_running(), so I had missed it: there's an hard-coded check for the state of dbus.service there, and it was missing the update to consider the new state. It is now fixed so I can drop the message. It's a pity to use the useful and accurate return error, but oh well.